Storage working group runtime changes

shamil-gadelshin opened the pull request:
## Implemented changes
- [Data directory module](https://github.com/Joystream/joystream/issues/566)
- [Data object storage registry module](https://github.com/Joystream/joystream/issues/565)
- [Data object type registry module](https://github.com/Joystream/joystream/issues/564)
- [Service discovery module](https://github.com/Joystream/joystream/issues/569)
- [Bureaucracy module](https://github.com/Joystream/joystream/issues/571)
    - implemented only: add method to enumerate workers - it is required to generate a random storage provider.
- [Runtime](https://github.com/Joystream/joystream/issues/573)
    - remove LookUpRoles
    - implement StorageProviderHelper - to allow picking a random provider


[Root issue is here](https://github.com/Joystream/joystream/issues/544).

shamil-gadelshin committed:
Update tests in the storage module

- Remove old tests.rs file
- Introduce a new tests module
- Move mock and all tests into the tests module

shamil-gadelshin committed:
Add bureacracy dependency and fix tests in the storage module

shamil-gadelshin committed:
Migrate data object type registry module with bureaucracy module

- change ensure_root() to the ensure_origin_is_active_leader()
- update tests
- add comments

shamil-gadelshin committed:
Refactor storage module

- remove unused trait ContentHasStorage
- move traits and delete traits.rs
- add comments
- remove incorrect TODOs

shamil-gadelshin committed:
Update data_object_storage_registry module

- remove unused events
- change ensure_signed to the ensure_worker_signed in the extrinsics
- update module tests

shamil-gadelshin committed:
Migrate  storage module to substrate errors

Modules affected:
- data object type registry
- data object storage registry

shamil-gadelshin committed:
Update comments for the data object storage registry module

- fix mocks

shamil-gadelshin committed:
Remove unused extrinsics

shamil-gadelshin committed:
Update storage::data_directory module

- add membership validation for the add_content()
- add StorageProviderHelper
- add runtime integration
- add some comments
- add signed storage provider check for accept_content() and reject_content().

shamil-gadelshin committed:
Add comments to the data_directory storage module.

shamil-gadelshin committed:
Restore the tests

shamil-gadelshin committed:
Refactor data_directory module

- remove unused code
- move BlockAndTime to the common module

shamil-gadelshin committed:
Refactor forum module: remove BlockchainTimestamp

Use BlockAndTime from the common module instead of duplicated BlockchainTimestamp.

shamil-gadelshin committed:
Add tests to the data_directory module

shamil-gadelshin committed:
Update service discovery module

- rename AccountInfoByAccountId to the AccountInfoByStorageProviderId
-add module comments
- modify tests
- modify runtime

shamil-gadelshin committed:
Migrate service directory to the substrate-style errors

shamil-gadelshin committed:
Update storage related versions in Cargo.toml’s

- set storage module version 2.0.0
- set service discovery module version 2.0.0
- update runtime to use these versions

shamil-gadelshin committed:
Add get_all_worker_ids() with tests to teh bureaucracy module

shamil-gadelshin committed:
Implement StorageProviderHelper in the runtime

shamil-gadelshin committed:
Rename test folder to tests in the runtime

shamil-gadelshin committed:
Implement StorageProviderHelper in the runtime

- implement StorageProviderHelper in the runtime
- add integration test

mnaamani commented on runtime-modules/bureaucracy/src/lib.rs:
```diff
@@ -975,6 +975,13 @@ impl<T: Trait<I>, I: Instance> Module<T, I> {
+    pub fn get_all_worker_ids() -> Vec<WorkerId<T>> {
+        <WorkerById<T, I>>::enumerate()
+            .map(|(worker_id, _)| worker_id)
+            .collect()
```
do we need to filter on 'active' workers ?

mnaamani replied:
So only selecting workers that are not unbonding

shamil-gadelshin replied:
In contrast with the content working group, we don't have 'unstaking worker stage' and we keep only active workers within the workers collection.

mnaamani commented on runtime-modules/bureaucracy/src/tests/mod.rs:
```diff
@@ -1529,3 +1528,25 @@ fn slash_worker_stake_fails_with_not_set_lead() {
+        expected_ids.sort();
+
+        let mut worker_ids = Bureaucracy1::get_all_worker_ids();
+        worker_ids.sort();
```
I'm assuming sorting is needed here because iterating over the map in get_all_worker_ids() is not guaranteed to have same order as inserting?

shamil-gadelshin replied:
Correct. Sorting will guarantee the element order.

mnaamani commented on runtime-modules/common/Cargo.toml:
```diff
@@ -1,6 +1,6 @@
 [package]
 name = 'substrate-common-module'
-version = '1.0.0'
+version = '1.0.1'
```
semantically you have added something to the api and not a bug-fix/patch so it make more sense to bump to `1.1.0`

shamil-gadelshin replied:
Agree. I'll fix it.

mnaamani commented on runtime-modules/common/src/lib.rs:
```diff
@@ -4,3 +4,27 @@
+    pub time: Moment,
+}
+
+/// Gathers current block and time information for the runtime
```
Can we add a comment here that explains if this function is used inside a `config()` at genesis the timestamp will be `0` because the timestamp is actually produced by validators.

shamil-gadelshin replied:
I copy your comments to the code comments.

mnaamani commented on runtime-modules/forum/src/lib.rs:
```diff
@@ -72,20 +73,12 @@ pub trait ForumUserRegistry<AccountId> {
-    block: BlockNumber,
-    time: Moment,
-}
-
```
@Lezek123 If you can update https://github.com/Joystream/joystream/pull/640 to include this refactoring.
We are dropping `BlockchainTimestamp` type and keeping only `BlockAndTime` struct which are identical.

mnaamani commented on runtime-modules/service-discovery/Cargo.toml:
```diff
@@ -56,15 +61,52 @@ rev = 'c37bb08535c49a12320af7facfd555ce05cce2e8'
 features = ['derive']
 package = 'parity-scale-codec'
-version = '1.0.0'
+version = '1.1.0'
```
is this required specifically or is it just a result of doing cargo-update ?

shamil-gadelshin replied:
I'm not sure how it happened. Whether it was changed somehow automatically or it was my mistake. Anyway, I downgrade it back to '1.0.0'.

mnaamani commented on Cargo.lock:
```diff
@@ -1614,7 +1614,7 @@ dependencies = [
 [[package]]
 name = "joystream-node-runtime"
-version = "6.14.0"
+version = "6.15.0"
```
Lets actually bump the runtime spec version from 15 to 16 and the version in Cargo.toml for the runtime crate to `6.16.0`

shamil-gadelshin replied:
Ok

mnaamani reviewed (changes requested):
I'm only halfway through review but I wanted to share small comments so far and for @Lezek123 to see the change in BlockAndTime

mnaamani commented on runtime-modules/service-discovery/src/lib.rs:
```diff
@@ -1,7 +1,192 @@
+            origin,
+            storage_provider_id: StorageProviderId<T>,
+            id: Vec<u8>,
+            lifetime: Option<T::BlockNumber>
```
I think since we are making changes, we should drop the user supplied lifetime. It doesn't make sense for it to be selected by the storage provider. And the runtime should just use the `default_lifetime()` for the ttl

shamil-gadelshin replied:
Ok. I'll fix it.

mnaamani commented on runtime-modules/service-discovery/src/lib.rs:
```diff
@@ -1,7 +1,192 @@
+        }
+
+        /// Sets bootstrap endpoints for the Colossus. Requires root privileges.
+        pub fn set_bootstrap_endpoints(origin, endpoints: Vec<Vec<u8>>) {
```
```suggestion
        pub fn set_bootstrap_endpoints(origin, endpoints: Vec<Url>) {
```

mnaamani commented on runtime-modules/storage/src/data_object_storage_registry.rs:
```diff
@@ -193,93 +238,9 @@ impl<T: Trait> Module<T> {
         <Relationships<T>>::insert(id, dosr);
         Self::deposit_event(RawEvent::DataObjectStorageRelationshipReadyUpdated(
-            id, true,
+            id, ready,
```
bug fix :+1:

mnaamani commented on runtime/Cargo.toml:
```diff
@@ -5,7 +5,7 @@ edition = '2018'
 # Follow convention: https://github.com/Joystream/substrate-runtime-joystream/issues/1
 # {Authoring}.{Spec}.{Impl} of the RuntimeVersion
-version = '6.14.0'
+version = '6.15.0'
```
```suggestion
version = '6.16.0'
```

mnaamani commented on runtime/src/lib.rs:
```diff
@@ -142,7 +145,7 @@ pub const VERSION: RuntimeVersion = RuntimeVersion {
     impl_name: create_runtime_str!("joystream-node"),
     authoring_version: 6,
-    spec_version: 14,
+    spec_version: 15,
```
```suggestion
    spec_version: 16,
```

mnaamani commented on runtime/src/tests/storage_integration.rs:
```diff
@@ -0,0 +1,42 @@
+fn storage_provider_helper_succeeds() {
+    initial_test_ext().execute_with(|| {
+
+		// Error - no workers.
```
formatting seems a little off here, but cargo-fmt isn't complaining, tab vs space indentation

shamil-gadelshin replied:
Added one line break. cargo fmt is still OK.

mnaamani reviewed (changes requested):
Looks good, we just need to bump spec version in `runtime/lib.rs` to 16

shamil-gadelshin committed:
Review fixes

- bump runtime version to the 6.16.0
- add comments
- restore parity-scale-codec version (1.0.0)
- change common module version to 1.1.0

shamil-gadelshin committed:
Merge branch 'nicaea' into storage_working_group_runtime_changes

# Conflicts:
#	Cargo.lock
#	runtime/Cargo.toml
#	runtime/src/lib.rs

shamil-gadelshin committed:
Fix formatting

shamil-gadelshin committed:
Remove user-defined lifetime from the service discovery module

- remove user-defined lifetime from the service discovery module (set_ipns_id).
